### DISCUSSION TITLE
Simpler String Interpolations in Template Literals

### DISCUSSION THREAD

**Discussion by wmstack** [DanielXMoore/Civet#1506] [Ideas]
Currently template literals are quite verbose requiring a minimum of 3 characters `${}` to interpolate an expression which I think is really unnecessary. For a single variable it could be reduced to `#variable` or `$variable` inside the template string, and for more complex expressions it can just be put directly between braces, like Rust for example: `{expression}`.

This would mean one has to escape regular braces but I think this is satisfactory because there are 3 ways to create strings.

This could be extended to allow `#variable.key` or `$var.key`, making object properties easier to interpolate without an extra layer of braces.

**Comment by edemaine** [comment]
I'm not sure I would like this personally; we would have to escape every `$` or `#` that wasn't an interpolation. Currently it's pretty hard to trigger an accidental interpolation because `${` is uncommon, but with just `$` it would be relatively common I think.

I could see adding this as a compiler flag, though. I guess it's similar to Bourne shell interpolation.

**Comment by STRd6** [comment]
It's a neat idea and potentially worth exploring but the primary challenge is that it would create an incompatibility with existing JS code. We do have some leeway with triple backtick and triple `"""` blocks but we also don't necessarily want them to behave very differently without a compelling reason.

**Comment by bbrk24** [comment]
I'll note that this isn't just a shell thing; Kotlin and PHP do it too, for example. Also, I escape `$`s in backtick-strings anyways, so that doesn't bother me.

As for other suggestions: maybe require a prefix, like `f"..."` in Python or `$"..."` in C#?

**Reply by edemaine** [reply]
Thanks for the context! I didn't know those languages had this too. 

Prefixes like that already have a meaning in Civet (template literals). Which you could actually use to implement something like this at the user level...

Anyway, it sounds like there's interest to add a compiler flag to allow this? Something like "shortInterpolations" perhaps? I imagine it would enable `$` in template strings, and if coffeeInterpolations is also on, `#` in double quoted strings.

**Reply by wmstack** [reply]
Yeah that would be nice, as well as the possibility of interpolating using braces `{}` without a prefix, like Rust with their format and println macros. My preference is interpolation in single quote strings `'` using the `#` symbol, but I don't mind as long as one type of string supports it.

**Reply by wmstack** [reply]
Well another language that has the syntax for interpolation like `"Hello, $world"` is Julia.

**Comment by STRd6** [comment]
I'm in favor of adding this as an option. We may even eventually consider making it the default option in conjunction with adding an esCompat option for migrating from TS/JS code.
